src/vumeter-qt/vumeter.cc

```diff
@@ -0,0 +1,437 @@
+
+    QFont font = p.font();
+    font.setPointSizeF(fminf(font_size_width, font_size_height));
+    p.setFont(font);
+
+    QPen pen = p.pen();
+    pen.setWidth(legend_line_width);
+    pen.setColor(text_color);
+    p.setPen(pen);
+
+    draw_vu_legend_db(p, 0, "0");
+    draw_vu_legend_db(p, -3, "-3");
+    draw_vu_legend_db(p, -6, "-6");
+    draw_vu_legend_db(p, -9, "-9");
+    draw_vu_legend_db(p, -12, "-12");
+    draw_vu_legend_db(p, -15, "-15");
+    draw_vu_legend_db(p, -18, "-18");
+    draw_vu_legend_db(p, -20, "-20");
+    draw_vu_legend_db(p, -25, "-25");
+    draw_vu_legend_db(p, -30, "-30");
+    draw_vu_legend_db(p, -35, "-35");
+    draw_vu_legend_db(p, -40, "-40");
+    draw_vu_legend_db(p, -50, "-50");
+    draw_vu_legend_db(p, -60, "-60");
+    draw_vu_legend_db(p, -VUMeterQt::db_range, "-inf");
+
+    pen.setColor(db_line_color);
+    p.setPen(pen);
+    for (int i = 0; i >= -60; i--)
+    {
+        if (i > -30)
+        {
+            draw_vu_legend_line(p, i);
+            draw_vu_legend_line(p, i - 0.5, 0.5);
+        }
+        else if (i > -40)
+        {
+            draw_vu_legend_line(p, i);
+        }
+        else if (i >= -60)
+        {
+            draw_vu_legend_line(p, i);
+            i -= 1;
+        }
+    }
+    draw_vu_legend_line(p, -VUMeterQt::db_range);
+}
+
+void VUMeterQtWidget::draw_vu_legend_line(QPainter &p, float db, float line_width_factor)
+{
+    float y = get_y_from_db(db);
+    if (db > -VUMeterQt::db_range) {
+        y += (legend_line_width / 2.0f);
+    } else {
+        y -= (legend_line_width / 2.0f);
+    }
+    float line_width = fclamp(legend_width * 0.25f, 1, 8);
+    p.drawLine(
+        QPointF(legend_width - line_width * line_width_factor - (legend_line_width / 2.0f), y),
+        QPointF(legend_width - (legend_line_width / 2.0f), y)
+    );
+    p.drawLine(
+        QPointF(width() - legend_width + (legend_line_width / 2.0f), y),
+        QPointF(width() - legend_width + (legend_line_width / 2.0f) + line_width * line_width_factor, y)
+    );
+}
+
+void VUMeterQtWidget::draw_vu_legend_db(QPainter &p, float db, const char *text)
+{
+    QFontMetricsF fm(p.font());
+    QSizeF text_size = fm.size(0, text);
+    float y = get_y_from_db(db);
+    float padding = fclamp(legend_width * 0.25f, 1, 8) * 1.5f;
+    p.drawText(QPointF(legend_width - text_size.width() - padding, y + (text_size.height()/4.0f)), text);
+    p.drawText(QPointF(width() - legend_width + padding, y + (text_size.height()/4.0f)), text);
+}
+
+void VUMeterQtWidget::draw_visualizer_peaks(QPainter &p)
+{
+    float bar_width = get_bar_width(nchannels);
+    float font_size_width = bar_width / 3.0f;
+    float font_size_height = vumeter_top_padding * 0.50f;
+
+    QFont font = p.font();
+    font.setPointSizeF(fminf(font_size_width, font_size_height));
+    p.setFont(font);
+
+    QPen pen = p.pen();
+    pen.setColor(text_color);
+    p.setPen(pen);
+
+    QFontMetricsF fm(p.font());
+    char text[10];
```

Comment:
while this is safe i'd prefer to see the `String` type used here with str functions for consistency with the rest of the codebase